[red-knot] inferred type, not Unknown, for undeclared paths

carljm opened the pull request:
After looking at more cases (for example, the case in the added test in this PR), I realized that our previous rule, "if a symbol has any declarations, use only declarations for its public type" is not adequate. Rather than using `Unknown` as fallback if the symbol is not declared in some paths, we need to use the inferred type as fallback in that case.

For the paths where the symbol _was_ declared, we know that any bindings must be assignable to the declared type in that path, so this won't change the overall declared type in those paths. But for paths where the symbol wasn't declared, this will give us a better type in place of `Unknown`.

MichaReiser commented:
> After looking at more cases (for example, the case in the added test in this PR), I realized that our previous rule, "if a symbol has any declarations, use only declarations for its public type" is not adequate. Rather than using Unknown as fallback if the symbol is not declared in some paths, we need to use the inferred type as fallback in that case.

Do you have some examples that you could share? It would be helpful if we look back to this PR in a few weeks and wonder why we changed it the way we did

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -173,26 +185,21 @@ type DeclaredTypeResult<'db> = Result<Type<'db>, (Type<'db>, Box<[Type<'db>]>)>;
-    })
-    .into_iter()
-    .chain(decl_types);
+    let mut all_types = undeclared_ty.into_iter().chain(decl_types);
```
Is it now possible that we emit one extra error if the `undeclared_ty` is not compatible with any of the declared ty?

carljm commented:
> Could you share some of those examples? It would be helpful if we look back to this PR in a few weeks and wonder why we changed it the way we did

I think the test I added summarizes the essential character of all the cases where this makes a difference; it's just the scenario where a name is declared in one path but not in another.
